File: SOFTWARE/APPMovilCanales/emv_nfc_plugin/ios/Classes/EmvCardReader.swift
import CoreNFC

@available(iOS 13.0, *)
class EmvCardReader {

    private let iso7816Tag: NFCISO7816Tag // The ISO 7816 tag for communication

    private let AID_VISA = "A0000000031010" // Visa Application Identifier
    private let PDOL_TAG_TTQ = "9F66" // Terminal Transaction Qualifiers
    private let PDOL_TAG_AMOUNT_AUTH = "9F02" // Amount, Authorized (Numeric)
    private let PDOL_TAG_AMOUNT_OTHER = "9F03" // Amount, Other (Numeric)
    private let PDOL_TAG_COUNTRY_CODE = "9F1A" // Terminal Country Code
    private let PDOL_TAG_TVR = "95" // Terminal Verification Results
    private let PDOL_TAG_CURRENCY_CODE = "5F2A" // Transaction Currency Code
    private let PDOL_TAG_DATE = "9A" // Transaction Date
    private let PDOL_TAG_TRANSACTION_TYPE = "9C" // Transaction Type
    private let PDOL_TAG_UNPREDICTABLE_NUMBER = "9F37" // Unpredictable Number

    init(iso7816Tag: NFCISO7816Tag) {
        self.iso7816Tag = iso7816Tag
    }

    func readEmvCard() async throws -> [String: String]? { // Marked as async and throws
        var cardData = [String: String]()

        let selectApdu = buildSelectAidApdu(aid: AID_VISA)
        let selectResponse: Data
        do {
            selectResponse = try await sendCommand(apdu: selectApdu, commandName: "SELECT AID")
        } catch {
            throw error // Re-throw the error to be handled by the caller
        }

        guard isSuccess(response: selectResponse) else {
            return nil
        }

        let pdolHex: String? = {
            guard let tlv1 = parseTlv(hexData: bytesToHex(selectResponse), context: "SELECT AID Response (Outer)"),
                  let fci = tlv1["6F"].flatMap({ parseTlv(hexData: $0, context: "FCI Template (6F)") }),
                  let a5 = fci["A5"].flatMap({ parseTlv(hexData: $0, context: "Proprietary Template (A5)") }) else {
                return nil
            }
            return a5["9F38"]
        }()

        let gpoApdu = buildGpoApdu(pdolHex: pdolHex)
        let gpoResponse: Data
        do {
            gpoResponse = try await sendCommand(apdu: gpoApdu, commandName: "GET PROCESSING OPTIONS")
        } catch {
            throw error // Re-throw the error
        }

        guard isSuccess(response: gpoResponse) else {
            return nil
        }

        let gpoHex = bytesToHex(gpoResponse)
        guard let tlvMainDict = parseTlv(hexData: gpoHex, context: "GPO Response (Main)") else {
            return nil
        }

        let template77: [String: String]
        if let parsedTemplate77 = tlvMainDict["77"].flatMap({ parseTlv(hexData: $0, context: "Template 77") }) {
            template77 = parsedTemplate77
        } else {
            template77 = tlvMainDict
        }

        if let track2 = template77["57"] {
            let parts = track2.split(separator: "D")
            if parts.count >= 2 {
                let pan = String(parts[0])
                let expiryData = String(parts[1])
                if expiryData.count >= 4 {
                    let expYY = String(expiryData.prefix(2))
                    let expMM = String(expiryData[expiryData.index(expiryData.startIndex, offsetBy: 2)..<expiryData.index(expiryData.startIndex, offsetBy: 4)])
                    cardData["PAN"] = pan
                    cardData["EXPIRY_DATE"] = "\(expMM)/\(expYY)"
                } else {
                    return nil
                }
            } else {
                return nil
            }
        } else {
            return nil // Track 2 data not found
        }

        return cardData
    }

    private func sendCommand(apdu: Data, commandName: String) async throws -> Data { // Marked as async and throws
        do {
            let (response, sw1, sw2) = try await iso7816Tag.sendCommand(apdu: NFCISO7816APDU(data: apdu)!)
            // Combine response data with SW1 and SW2 status bytes
            var fullResponse = response
            fullResponse.append(sw1)
            fullResponse.append(sw2)
            return fullResponse
        } catch let error as NFCReaderError {
            throw error
        } catch {
            throw error
        }
    }

    private func isSuccess(response: Data) -> Bool {
        guard response.count >= 2 else {
            return false
        }
        let sw1 = response[response.count - 2]
        let sw2 = response[response.count - 1]
        let success = (sw1 == 0x90 && sw2 == 0x00)
        return success
    }

    private func hexToBytes(hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            if let byte = UInt8(hex[index..<nextIndex], radix: 16) {
                data.append(byte)
            } else {
                // Handle error, e.g., return empty data or throw
            }
            index = nextIndex
        }
        return data
    }

    private func bytesToHex(_ bytes: Data) -> String {
        return bytes.map { String(format: "%02X", $0) }.joined()
    }

    private func parseTlv(hexData: String, context: String = "Unknown") -> [String: String]? {
        var tlvMap = [String: String]()
        var currentIndex = hexData.startIndex

        while currentIndex < hexData.endIndex {
            var tagBytesCount = 0
            var tagEndIndex = currentIndex
            var tagHex = ""

            guard let firstTagByteRange = hexData.range(of: hexData[currentIndex..<hexData.endIndex].prefix(2)),
                  let firstTagByte = UInt8(hexData[firstTagByteRange], radix: 16) else {
                return nil
            }
            tagEndIndex = hexData.index(currentIndex, offsetBy: 2)
            tagHex = String(hexData[currentIndex..<tagEndIndex])
            tagBytesCount += 1

            if (firstTagByte & 0x1F) == 0x1F { // Multi-byte tag
                while tagEndIndex < hexData.endIndex {
                    guard let nextTagByteRange = hexData.range(of: hexData[tagEndIndex..<hexData.endIndex].prefix(2)),
                          let nextTagByte = UInt8(hexData[nextTagByteRange], radix: 16) else {
                        return nil
                    }
                    tagEndIndex = hexData.index(tagEndIndex, offsetBy: 2)
                    let startOfLastByte = hexData.index(tagEndIndex, offsetBy: -2)
                    tagHex += String(hexData[startOfLastByte..<tagEndIndex]) // Append the new byte
                    tagBytesCount += 1
                    if (nextTagByte & 0x80) == 0 { break } // Last byte of multi-byte tag
                }
            }
            let tag = tagHex.uppercased()
            currentIndex = tagEndIndex
            let tagStartIndex = hexData.index(currentIndex, offsetBy: -tagHex.count)

            guard currentIndex < hexData.endIndex else {
                return nil
            }
            guard let lengthByteRange = hexData.range(of: hexData[currentIndex..<hexData.endIndex].prefix(2)),
                  var lengthByte = UInt8(hexData[lengthByteRange], radix: 16) else {
                return nil
            }
            currentIndex = hexData.index(currentIndex, offsetBy: 2)

            var valueLength: Int
            if (lengthByte & 0x80) == 0x80 { // Length is encoded in multiple bytes
                let numLengthBytes = Int(lengthByte & 0x7F)
                guard numLengthBytes > 0 else {
                    return nil
                }
                guard hexData.endIndex.encodedOffset - currentIndex.encodedOffset >= numLengthBytes * 2 else {
                    return nil
                }
                let actualLengthHex = String(hexData[currentIndex..<hexData.index(currentIndex, offsetBy: numLengthBytes * 2)])
                guard let parsedLength = Int(actualLengthHex, radix: 16) else {
                    return nil
                }
                valueLength = parsedLength
                currentIndex = hexData.index(currentIndex, offsetBy: numLengthBytes * 2)
            } else {
                valueLength = Int(lengthByte)
            }

            guard hexData.endIndex.encodedOffset - currentIndex.encodedOffset >= valueLength * 2 else {
                return nil
            }
            let valueEndIndex = hexData.index(currentIndex, offsetBy: valueLength * 2)
            let value = String(hexData[currentIndex..<valueEndIndex]).uppercased()
            currentIndex = valueEndIndex

            tlvMap[tag] = value
        }
        return tlvMap
    }

    /// Construye un comando APDU para seleccionar una aplicación (AID) en la tarjeta.
    ///
    /// El comando SELECT (CLA: 00, INS: A4) se utiliza para seleccionar una aplicación específica.
    /// Formato: [00 A4 04 00 Lc AID 00]
    /// - CLA (00): Instrucción estándar.
    /// - INS (A4): Comando SELECT.
    /// - P1 (04): Seleccionar por nombre (AID).
    /// - P2 (00): Primer o único resultado.
    /// - Lc: Longitud del AID.
    /// - AID: Identificador de la aplicación.
    /// - Le (00): Esperar respuesta completa.
    ///
    /// - Parameter aid: El Application Identifier (AID) como una cadena hexadecimal.
    /// - Returns: Un objeto `Data` que representa el APDU completo.
    private func buildSelectAidApdu(aid: String) -> Data {
        let aidBytes = hexToBytes(hex: aid)
        let aidLength = UInt8(aidBytes.count)

        var apdu = Data()
        apdu.append(0x00) // CLA
        apdu.append(0xA4) // INS
        apdu.append(0x04) // P1
        apdu.append(0x00) // P2
        apdu.append(aidLength) // Lc
        apdu.append(aidBytes) // Datos (AID)
        apdu.append(0x00) // Le
        return apdu
    }

    /// Construye el comando APDU para el GPO (Get Processing Options).
    ///
    /// Este comando inicia el procesamiento de la transacción. Utiliza el Tag 83 para encapsular
    /// los datos del PDOL, que son información que el terminal proporciona a la tarjeta.
    ///
    /// Formato: [80 A8 00 00 Lc 83 XX ...]
    /// - Si no se recibe un PDOL, se envía un campo de datos "8300" (GPO sin PDOL).
    ///
    /// - Parameter pdolHex: La cadena hexadecimal opcional que representa los tags y longitudes del PDOL.
    /// - Returns: Un objeto `Data` que representa el APDU de GPO.
    private func buildGpoApdu(pdolHex: String?) -> Data {
        let commandHeader = "80A80000"
        var dataField = "83" // Tag para el PDOL en el GPO
        var pdolValueHex = ""

        if let pdol = pdolHex, !pdol.isEmpty {
            var currentIndex = pdol.startIndex

            while currentIndex < pdol.endIndex {
                var tagEndIndex = pdol.index(currentIndex, offsetBy: 2)
                var tagHex = String(pdol[currentIndex..<tagEndIndex])

                if (Int(tagHex, radix: 16)! & 0x1F) == 0x1F {
                    guard pdol.index(tagEndIndex, offsetBy: 2) <= pdol.endIndex else {
                        break
                    }
                    tagEndIndex = pdol.index(tagEndIndex, offsetBy: 2)
                    tagHex = String(pdol[currentIndex..<tagEndIndex])
                }
                currentIndex = tagEndIndex

                // Extract Length
                guard pdol.index(currentIndex, offsetBy: 2) <= pdol.endIndex else {
                    break
                }
                let lengthEndIndex = pdol.index(currentIndex, offsetBy: 2)
                let len = Int(String(pdol[currentIndex..<lengthEndIndex]), radix: 16)!
                currentIndex = lengthEndIndex

                // Generate simulated data based on the tag
                let simulatedData: String
                switch tagHex.uppercased() {
                    case PDOL_TAG_TTQ: simulatedData = "20000000" // Terminal Transaction Qualifiers
                    case PDOL_TAG_AMOUNT_AUTH: simulatedData = String(repeating: "0", count: len * 2) // Amount, Authorized (Numeric)
                    case PDOL_TAG_AMOUNT_OTHER: simulatedData = String(repeating: "0", count: len * 2) // Amount, Other (Numeric)
                    case PDOL_TAG_COUNTRY_CODE: simulatedData = "0604" // Perú (ISO 3166-1 numeric for Peru)
                    case PDOL_TAG_TVR: simulatedData = String(repeating: "0", count: len * 2) // Terminal Verification Results
                    case PDOL_TAG_CURRENCY_CODE: simulatedData = "0604" // PEN (ISO 4217 numeric for Peruvian Sol)
                    case PDOL_TAG_DATE:
                        let formatter = DateFormatter()
                        formatter.dateFormat = "yyMMdd"
                        simulatedData = formatter.string(from: Date())
                    case PDOL_TAG_TRANSACTION_TYPE: simulatedData = "00" // Purchase
                    case PDOL_TAG_UNPREDICTABLE_NUMBER:
                        // Generate a random 4-byte (8 hex char) unpredictable number
                        simulatedData = String(format: "%08X", Int.random(in: 0...Int.max))
                    default:
                        simulatedData = String(repeating: "0", count: len * 2)
                }
                pdolValueHex += simulatedData.prefix(len * 2)
            }
        } else {
            pdolValueHex = ""
        }

        dataField += String(format: "%02X", pdolValueHex.count / 2) + pdolValueHex

        let dataLength = String(format: "%02X", dataField.count / 2)

        let finalApduHex = commandHeader + dataLength + dataField
        return hexToBytes(hex: finalApduHex)
    }
}

File: SOFTWARE/APPMovilCanales/emv_nfc_plugin/android/src/main/kotlin/com/cajatacna/emv_nfc_plugin/EmvCardReader.kt
package com.cajatacna.emv_nfc_plugin

import android.nfc.tech.IsoDep
import java.nio.ByteBuffer
import java.util.*
import kotlin.collections.LinkedHashMap
import java.time.format.DateTimeFormatter
import java.time.LocalDate
import kotlin.random.Random

class EmvCardReader(private val isoDep: IsoDep) {

    private val HEX_ARRAY = "0123456789ABCDEF".toCharArray()
    private val AID_VISA = "A0000000031010"
    private val PDOL_TAG_TTQ = "9F66" // Terminal Transaction Qualifiers
    private val PDOL_TAG_AMOUNT_AUTH = "9F02" // Amount, Authorized (Numeric)
    private val PDOL_TAG_AMOUNT_OTHER = "9F03" // Amount, Other (Numeric)
    private val PDOL_TAG_COUNTRY_CODE = "9F1A" // Terminal Country Code
    private val PDOL_TAG_TVR = "95" // Terminal Verification Results
    private val PDOL_TAG_CURRENCY_CODE = "5F2A" // Transaction Currency Code
    private val PDOL_TAG_DATE = "9A" // Transaction Date
    private val PDOL_TAG_TRANSACTION_TYPE = "9C" // Transaction Type
    private val PDOL_TAG_UNPREDICTABLE_NUMBER = "9F37" // Unpredictable Number

    fun readEmvCard(): Map<String, String>? {
        val cardData = LinkedHashMap<String, String>()

        try {
            val selectApdu = buildSelectAidApdu(AID_VISA)
            val selectResponse = isoDep.transceive(selectApdu)
            if (!isSuccess(selectResponse))
                return null

            val pdol = runCatching {
                val tlv1 = parseTlv(bytesToHex(selectResponse))
                val fci = tlv1["6F"]?.let { parseTlv(it) }
                val a5 = fci?.get("A5")?.let { parseTlv(it) }
                a5?.get("9F38")
            }.getOrNull()

            val gpoApdu = buildGpoApdu(pdol)
            val gpoResponse = isoDep.transceive(gpoApdu)
            if (!isSuccess(gpoResponse))
                return null

            val gpoHex = bytesToHex(gpoResponse)
            val tlvMain = parseTlv(gpoHex)
            val template77 = tlvMain["77"]?.let { parseTlv(it) } ?: tlvMain

            template77["57"]?.let { track2 ->
                val parts = track2.split("D")
                if (parts.size >= 2) {
                    val pan = parts[0]
                    val expYY = parts[1].substring(0, 2)
                    val expMM = parts[1].substring(2, 4)
                    cardData["PAN"] = pan
                    cardData["EXPIRY_DATE"] = "$expMM/$expYY"
                }
            } ?: run {
                return null
            }

            return cardData
        } catch (e: Exception) {
            return null
        }
    }

    /**
     * Construye un comando APDU para seleccionar una aplicación (AID) en la tarjeta.
     *
     * El comando SELECT (CLA: 00, INS: A4) se utiliza para seleccionar una aplicación
     * específica dentro de una tarjeta inteligente EMV. Este método arma el APDU en formato:
     * [00 A4 04 00 Lc AID 00], donde:
     * - CLA = 00 (instrucción estándar)
     * - INS = A4 (comando SELECT)
     * - P1 = 04 (seleccionar por nombre)
     * - P2 = 00 (primer o único resultado)
     * - Lc = longitud del AID
     * - AID = identificador de la aplicación en formato hexadecimal
     * - Le = 00 (esperar respuesta completa)
     *
     * @param aid AID (Application Identifier) como cadena hexadecimal (por ejemplo, "A0000000031010" para Visa).
     * @return Arreglo de bytes representando el APDU listo para ser enviado mediante NFC a la tarjeta.
     */
    private fun buildSelectAidApdu(aid: String): ByteArray {
        val aidBytes = hexToBytes(aid)
        val aidLength = aidBytes.size
        val apdu = ByteBuffer.allocate(6 + aidLength)
        apdu.put(0x00.toByte())
        apdu.put(0xA4.toByte())
        apdu.put(0x04.toByte())
        apdu.put(0x00.toByte())
        apdu.put(aidLength.toByte())
        apdu.put(aidBytes)
        apdu.put(0x00.toByte())
        return apdu.array()
    }

    /**
     * Construye el comando APDU para el GPO (Get Processing Options) de una tarjeta EMV.
     *
     * Este comando sigue la estructura definida por el estándar EMV para iniciar el procesamiento de la aplicación
     * seleccionada. Utiliza el Tag 83 para encapsular los datos del PDOL (Processing Data Object List), que contiene
     * información que el terminal proporciona a la tarjeta para iniciar la transacción.
     *
     * El comando tiene el siguiente formato:
     * [80 A8 00 00 Lc 83 XX ...] donde:
     * - CLA = 0x80 (proprietary class)
     * - INS = 0xA8 (Get Processing Options)
     * - P1, P2 = 0x00
     * - Lc = longitud total del campo de datos
     * - 83 = tag EMV para indicar el uso de PDOL
     * - XX = longitud del contenido del PDOL (en bytes)
     * - Datos = valores simulados para cada tag del PDOL requerido por la tarjeta
     *
     * Si no se recibe un PDOL (es decir, es null o vacío), se envía un campo de datos "8300", que significa GPO sin PDOL.
     *
     * @param pdolHex Cadena hexadecimal que representa los tags y longitudes del PDOL devuelto por la tarjeta en el SELECT AID.
     * @return APDU listo para ser enviado mediante NFC para ejecutar el comando GPO.
     */
    private fun buildGpoApdu(pdolHex: String?): ByteArray {
        val commandHeader = "80A80000" // CLA INS P1 P2
        var dataField = "83" // Tag para el PDOL en el GPO
        var pdolValueHex = "" // Contenido de datos del PDOL

        if (pdolHex.isNullOrEmpty()) {
            pdolValueHex = "00" // Longitud 00 para un PDOL vacío (8300)
        } else {
            var i = 0
            while (i < pdolHex.length) {
                var tagHex = pdolHex.substring(i, i + 2)
                var tagLenBytes = 1

                if ((tagHex.toInt(16) and 0x1F) == 0x1F) {
                    tagLenBytes = 2
                    tagHex = pdolHex.substring(i, minOf(i + 4, pdolHex.length))
                }
                i += (tagLenBytes * 2)

                if (i + 2 > pdolHex.length) {
                    break
                }

                val len = pdolHex.substring(i, i + 2).toInt(16)
                i += 2

                val simulatedData = when (tagHex.uppercase(Locale.ROOT)) {
                    PDOL_TAG_TTQ -> "20000000" // 4 bytes
                    PDOL_TAG_AMOUNT_AUTH -> "000000000000" // 6 bytes
                    PDOL_TAG_AMOUNT_OTHER -> "000000000000" // 6 bytes
                    PDOL_TAG_COUNTRY_CODE -> "0604" // 2 bytes (Perú)
                    PDOL_TAG_TVR -> "0000000000" // 5 bytes
                    PDOL_TAG_CURRENCY_CODE -> "0604" // 2 bytes (PEN)
                    PDOL_TAG_DATE -> LocalDate.now().format(DateTimeFormatter.ofPattern("yyMMdd")) // 3 bytes
                    PDOL_TAG_TRANSACTION_TYPE -> "00" // 1 byte (Purchase)
                    PDOL_TAG_UNPREDICTABLE_NUMBER -> String.format("%08X", Random.nextInt()) // 4 bytes
                    else -> "0".repeat(len * 2)
                }

                pdolValueHex += simulatedData.substring(0, minOf(simulatedData.length, len * 2)).padEnd(len * 2, '0')
            }
        }

        dataField += String.format("%02X", (pdolValueHex.length / 2)) + pdolValueHex
        val dataLength = String.format("%02X", (dataField.length / 2))

        return hexToBytes("$commandHeader$dataLength$dataField")
    }

    /**
     * Verifica si la respuesta APDU indica éxito.
     *
     * En el protocolo ISO/IEC 7816-4, una respuesta exitosa de una tarjeta (Smart Card o EMV) finaliza con los bytes
     * de estado 0x9000 (SW1=0x90, SW2=0x00). Este método extrae los dos últimos bytes de la respuesta para validar ese estado.
     *
     * @param response Arreglo de bytes devuelto por la tarjeta como respuesta a un comando APDU.
     * @return `true` si la respuesta termina con 0x9000 (éxito), `false` en cualquier otro caso.
     */
    private fun isSuccess(response: ByteArray): Boolean {
        if (response.size < 2) return false
        val sw1 = response[response.size - 2].toInt() and 0xFF
        val sw2 = response[response.size - 1].toInt() and 0xFF
        return sw1 == 0x90 && sw2 == 0x00
    }

    /**
     * Convierte una cadena hexadecimal (por ejemplo: "A0000000031010") en un array de bytes.
     *
     * Este método se usa para transformar comandos APDU representados como texto hexadecimal
     * en datos binarios reales que pueden ser enviados a través de NFC a una tarjeta.
     *
     * Por ejemplo, "A0000000031010" se convierte en:
     * [0xA0, 0x00, 0x00, 0x00, 0x03, 0x10, 0x10]
     *
     * @param hex Cadena en formato hexadecimal (sin espacios).
     * @return ByteArray representando los bytes reales.
     */
    private fun hexToBytes(hex: String): ByteArray {
        val len = hex.length
        val data = ByteArray(len / 2)
        for (i in 0 until len step 2) {
            data[i / 2] = ((Character.digit(hex[i], 16) shl 4) + Character.digit(hex[i + 1], 16)).toByte()
        }
        return data
    }

    /**
     * Convierte un arreglo de bytes en una cadena hexadecimal.
     *
     * Ejemplo: [0xA0, 0x00, 0x10] → "A00010"
     *
     * Se usa para mostrar de forma legible los comandos APDU o las respuestas de la tarjeta.
     */
    private fun bytesToHex(bytes: ByteArray): String {
        val hexChars = CharArray(bytes.size * 2)
        for (i in bytes.indices) {
            val v = bytes[i].toInt() and 0xFF
            hexChars[i * 2] = HEX_ARRAY[v ushr 4]
            hexChars[i * 2 + 1] = HEX_ARRAY[v and 0x0F]
        }
        return String(hexChars)
    }

    /**
     * Parsea una cadena hexadecimal codificada en formato TLV (Tag-Length-Value).
     *
     * Este método analiza una estructura TLV anidada o plana a partir de una cadena
     * hexadecimal, extrayendo cada campo con su respectivo tag, longitud y valor,
     * y los almacena en un mapa (LinkedHashMap) preservando el orden.
     *
     * Soporta tags de uno o varios bytes, longitudes codificadas en uno o más bytes
     * (formato BER-TLV), y valida que los datos no excedan los límites del string.
     *
     * @param hexData Cadena hexadecimal a analizar (por ejemplo, "6F1A8407A0000000031010A50F500B564953412044454249544F").
     * @return Mapa con los tags y sus valores hexadecimales correspondientes (ambos en mayúsculas).
     */
    private fun parseTlv(hexData: String): Map<String, String> {
        val tlvMap = LinkedHashMap<String, String>()
        var i = 0
        while (i < hexData.length - 1) {
            var tag: String
            var tagBytesCount = 0

            if (i + 2 > hexData.length) break
            val firstTagByte = hexData.substring(i, i + 2).toInt(16)
            i += 2
            tagBytesCount = 1

            if ((firstTagByte and 0x1F) == 0x1F) {
                while (i + 2 <= hexData.length) {
                    val nextTagByte = hexData.substring(i, i + 2).toInt(16)
                    i += 2
                    tagBytesCount++
                    if ((nextTagByte and 0x80) == 0) break
                }
            }
            tag = hexData.substring(i - (tagBytesCount * 2), i)

            if (i + 2 > hexData.length) break
            var lengthByte = hexData.substring(i, i + 2).toInt(16)
            i += 2

            var valueLength: Int
            if (lengthByte and 0x80 == 0x80) {
                val numLengthBytes = lengthByte and 0x7F
                if (i + (numLengthBytes * 2) > hexData.length) break
                val actualLengthHex = hexData.substring(i, i + (numLengthBytes * 2))
                valueLength = actualLengthHex.toInt(16)
                i += (numLengthBytes * 2)
            } else {
                valueLength = lengthByte
            }

            if (i + (valueLength * 2) > hexData.length) break
            val value = hexData.substring(i, i + (valueLength * 2))
            i += (valueLength * 2)

            tlvMap[tag.uppercase(Locale.ROOT)] = value.uppercase(Locale.ROOT)
        }
        return tlvMap
    }
}